### ISSUE TITLE
sqlite-net ARM processor

### ISSUE THREAD

**Issue by beerygaz** [praeclarum/sqlite-net#578]
I'm an inexperienced hobby programmer, so forgive my ignorance.

I'm writing a .net application that needs to run on Windows and Linux Mono (bot i386 and ARM processors).  I started using Syste.Data.SQLite and EF6 but that failed me on Mono, especially on the ARM platform.

Research suggested I use sqlite-net, and I love the simplicity and ease of use, but I still can't get it to work on the ARM Mono platform.  

Besides sqlit-net, I'm not sure which other Nuget packages are necessary for this multi-platform (single executable) deployment. 

The error I'm experiencing on the ARM Mono version is:
System.TypeInitializationException: An exception was thrown by the type initializer for SQLite.SQLiteConnection ---> System.DllNotFoundException: e_sqlite3

It seems to be looking for "libe_sqlite3.so"

I'd welcome some advice please?

**Comment by lubos** [comment]
libe_sqlite3.so is pre-compiled for x64. This means it won't work on ARM out of the box.

The workaround would be compile sqlite3.c yourself for ARM, then replace libe_sqlite3.so for your compiled build.

**Comment by beerygaz** [comment]
Thanks lubos.  Once complied, would I place the library in the /arm subdirectory like the other /x86 /x64 versions at present?

**Comment by lubos** [comment]
No. /x86 and /x64 sub-folders are for Windows only. On Linux, you need to have `libe_sqlite3.so` in the base directory just like most of your .NET assemblies.

**Comment by beerygaz** [comment]
I was afraid you'd say that.  My issue is I have no idea what architecture my application is being deployed to, so don't know which library to deploy.  Ideally I need to deploy all of the various assemblies and have the application select the most appropriate for the underlying OS and processor (Windows vs Mono, i386 vs ARM, etc).  I'm afraid I don't know where to start in trying to achieve this.  

Until now, I've been lucky that all of my applications complied against and x86 CPU have worked on Windows and Mono without issue.

**Comment by lubos** [comment]
I've discussed this with @ericsink who has made the underlying library SQLitePCL.raw and he has made serious effort into trying to make Linux load binary from sub-folders based on the platform but it became too complicated.

The only way would be to compile `SQLite3.c` on target machine during deployment. This means you wouldn't distribute any binary on Linux and your package could stay noarch. Or simply prepare multiple downloads - one for each platform.

**Comment by ericsink** [comment]
FWIW, I do expect this is going to work with .NET Core (as opposed to Mono).  In fact, I've just published a prerelease testing build of `SQLitePCLRaw.lib.e_sqlite3.linux` with ARM support included:

https://www.nuget.org/packages/SQLitePCLRaw.lib.e_sqlite3.linux/1.1.8-pre20170717084758

**Comment by praeclarum** [comment]
I still need to test this on an RPi -- hope it works!

**Comment by ericsink** [comment]
I have received feedback that my test package works (on .NET Core).  This will be part of the soon-to-be-released SQLitePCL.raw 1.1.8.
